packages/node_modules/nav-frontend-tabs/src/tabs.tsx

```diff
@@ -0,0 +1,108 @@
+import * as React from 'react';
+import * as cn from 'classnames';
+import { guid, omit } from 'nav-frontend-js-utils';
+
+import Tab, {TabProps} from './tab';
+
+import 'nav-frontend-tabs-style';
+
+const tabsCls = (props) => cn('nav-frontend-tabs', props.className, {
+    'nav-frontend-tabs--kompakt': props.kompakt
+});
+
+export interface TabsProps {
+    tabs: TabProps[];
+    onChange: (index: number) => void;
+    kompakt?: boolean;
+}
+
+export interface TabsState {
+    aktivTab: number;
```

Comment:
Trenger vi `state` her? Er ikke det allerede definert via `TabProps`?

Reply:
Default verdi til `state.aktivTab` blir gitt av props som overskrives ved endring. Blir litt som med stegindikatoren at `onChange` krever en intern state. Men også her går det an å legge til en pure versjon senere uten state.

Reply:
Hvordan blir `state.aktivTab` oppdaterert om man endrer på propsene egentlig? Eneste stedet hvor den utledes fra props er i konstruktøren, og den vil ikke nødvendigvis kjøre om man endrer på props til `Tabs`-komponenten.

Etter det jeg kan se på brukes brukes `state.aktivTab` bare i `renderTabs` for å sjekk om ett gitt steg skal være aktivt, men det vet jo hver enkelt `Tab` basert på sine props. Så ser ikke helt behovet her.

Reply:
Nei, det stemmer. Tenker kanskje at manuell kontroll over aktiv tab (ved å manipulere props underveis) hører mer naturlig hjemme i en pure versjon av komponenten, som kan legges til senere.

Reply:
pure-komponenten burde jo være grunnlaget for en "ikke-pure"-komponent. Så pure-komponenten burde vel bli lagt til først imho. ;) 

Om ikke-pure komponenten skal legges til først bør det tas en runde på api-et til komponenten. Per i dag synes jeg api-et signaliserer at komponenten er pure, mens i realiteten er den ikke det. Og det tror jeg fort kan by på forvirring down-the-line.

Reply:
Jau, det blir sikkert mest ryddig slik :) Er det `aktiv`-prop'en du tenker kan bli forvirrende i APIet? Hva tenker du om noe sånt i så fall?
```JSX
<Tabs 
    tabs={[
        { label: 'Første side', defaultAktiv: true },
        { label: 'Andre side' },
        { label: 'Tredje side' }
    ]}
    onChange={this.handleChange}
/>
```
```JSX
<TabsPure 
    tabs={[
        { label: 'Første side', aktiv: true },
        { label: 'Andre side' },
        { label: 'Tredje side' }
    ]}
/>
```

Reply:
Tenker prop'ene til `Tabs` kunne vært noe alà: 
```
interface TabsPureProps {
  tabs: TabProps[];
  onChange: (index: number) => void;
  kompakt?: boolean;
}
interface TabsPureState = {}; 
```
```
interface TabsProps {
  tabs: TabProps[];
  kompakt?: boolean;
  defaultAktiv: number;
  // Kanskje onChange her også? Men da optional?
}
interface TabsState {
  aktivTab: number;
}
```
Eller noe sånt..

Reply:
Hm, mulig jeg som har misforstått, men var ikke noe av vitsen med pure-komponentene at de skulle være stateless? Og i så fall vil vel `onChange` på `TabsPure` bli litt rart? 

Har skisset opp noe slikt:
```JSX
interface TabsPureProps {
    tabs: TabProps[];
    kompakt?: boolean;
}
```
```JSX
interface TabsProps extends TabsPureProps {
    onChange: (index: number) => void;
}

interface TabsState {
    aktivTab: number;
}
```
```JSX
interface TabProps {
    label?: React.ReactNode | React.ReactChild | React.ReactChildren;
    defaultAktiv?: boolean;
    aktiv?: boolean;
    onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
    linkCreator?: Function;
    disabled?: boolean;
}
```
Men det å flytte `defaultAktiv` til `TabsProps` kan sikkert være lurt.

Reply:
Man må nok ha `onChange` på pure-komponenten. Hvordan skal ellers de som bruker komponenten få vite at noen ønsket å bytte tab?

For ikke-pure-komponenten kan `onChange` være optional, siden der vil det ikke være nødvendig for konsumenten å vite at noen ønsket å bytte, da dette håndteres av komponenten selv. Eneste grunnen til at onChange kanskje skal være med der er om man skal godta andre side-effekter av at man bytter en tab.

Reply:
Nja, for pure-komponenten tenker jeg dette heller gjøres ved å ha en custom `onClick` handler på selve tabsene.

Merk at denne komponenten ikke innholder noen paneler (heller ikke i ikke-pure versjon). Så switching av paneler ved navigering mellom tabsene må uansett håndteres av konsumenten, som igjen gjør at `onChange` på ikke-pure versjon blir påkrevd.

Reply:
Det vil også fungere, men krever mer boilerplate for å oppnå akkurat samme funksjonalitet. 
`Tabs` kan fortsatt være stateless selvom den håndterer `onChange`, og tenkte det var lettere å forholde seg til en `onChange` fremfor å måtte sette det på hver enkelt `Tab`. 
Men effekten av det vil være akkurat den samme, så har ikke så mye å si for min del. 

Aaah, jeg har nok tenkt at panelene også er med her. Men det er forsåvidt bra, siden `Tabs` komponenten blir mer fleksibel om man ikke knytter innholdet inn i komponenten. Men ja, da blir  `onChange` påkrevd uansett, og er nesten så tror vi kunne ha sluppet unna med å bare ha en pure-versjon av komponenten.